src/main/java/fi/aalto/cs/apluscourses/intellij/model/ExercisesUpdater.java

```diff
@@ -121,20 +125,44 @@ private void addExercises(@NotNull List<ExerciseGroup> exerciseGroups,
                             @NotNull Progress progress) throws IOException {
     var course = courseProject.getCourse();
     var dataSource = course.getExerciseDataSource();
-    progress.incrementMaxValue(points.getExercisesCount());
+    var selectedStudent = courseProject.getSelectedStudent();
+    var exercisesTree = new ExercisesTree(exerciseGroups, selectedStudent);
+    for (var exerciseGroup : exerciseGroups) {
+
+      if (courseProject.isLazyLoaded(exerciseGroup.getId())) {
+        for (var exerciseId : points.getExercises(exerciseGroup.getId())) {
+          if (Thread.interrupted()) {
+            return;
+          }
+          var exercise = dataSource.getExercise(exerciseId, points, course.getTutorials(),
+              authentication, ZonedDateTime.now().minusDays(7));
+          exerciseGroup.addExercise(exercise);
+
+          progress.increment();
+
+          addSubmissionResults(exercise, points, authentication, progress);
+
+          courseProject.setExerciseTree(exercisesTree);
```

Comment:
This is set multiple times because it is in the loop. Intended? Anyway, it shouldn't.

Reply:
this way you can see the results as they load in